src/libaktualizr/uptane/fetcher.cc

```diff
@@ -49,32 +49,54 @@ static size_t DownloadHandler(char* contents, size_t size, size_t nmemb, void* u
     auto event = std::make_shared<event::DownloadProgressReport>(ds->target, "Downloading", calculated);
     (*(ds->events_channel))(event);
   }
+  if (ds->fetcher->isPaused()) {
+    return written_size + 1;
+  }
   return written_size;
 }
 
+void Fetcher::setPause(bool pause) {
+  if (pause) {
+    pause_mutex_.lock();
+  } else {
+    pause_mutex_.unlock();
+  }
+  pause_ = pause;
+}
+
 bool Fetcher::fetchVerifyTarget(const Target& target) {
   bool result = false;
   try {
     if (!target.IsOstree()) {
       DownloadMetaStruct ds(target, events_channel);
+      ds.fetcher = this;
       std::unique_ptr<StorageTargetWHandle> fhandle =
           storage->allocateTargetFile(false, target.filename(), static_cast<size_t>(target.length()));
       ds.fhandle = fhandle.get();
-      ds.downloaded_length = 0;
+      ds.downloaded_length = fhandle->getWrittenSize();
 
       if (target.hashes().empty()) {
         throw Exception("image", "No hash defined for the target");
       }
-
-      HttpResponse response =
-          http->download(config.uptane.repo_server + "/targets/" + target.filename(), DownloadHandler, &ds);
-      if (!response.isOk()) {
-        fhandle->wabort();
-        if (response.curl_code == CURLE_WRITE_ERROR) {
-          throw OversizedTarget(target.filename());
+      bool retry = false;
+      do {
+        retry = false;
+        HttpResponse response = http->download(config.uptane.repo_server + "/targets/" + target.filename(),
+                                               DownloadHandler, &ds, fhandle->getWrittenSize());
+        if (!response.isOk() && !pause_) {
+          fhandle->wabort();
+          if (response.curl_code == CURLE_WRITE_ERROR) {
+            throw OversizedTarget(target.filename());
+          }
+          throw Exception("image", "Could not download file, error: " + response.error_message);
         }
-        throw Exception("image", "Could not download file, error: " + response.error_message);
-      }
+        if (pause_) {
+          pause_mutex_.lock();
+          pause_mutex_.unlock();
```

Comment:
Why?

Reply:
We should unlock this mutex after another thread freed it because second time we will get  a dead lock. We don't use this mutex for data synchronization, only for waiting while it is blocked in another thread.